lang/java/avro/src/main/java/org/apache/avro/file/DataFileStream.java

```diff
@@ -97,18 +97,30 @@ protected DataFileStream(DatumReader<D> reader) throws IOException {
     this.reader = reader;
   }
 
-  /** Initialize the stream by reading from its head. */
-  void initialize(InputStream in) throws IOException {
-    this.header = new Header();
-    this.vin = DecoderFactory.get().binaryDecoder(in, vin);
+  byte[] readMagic() throws IOException {
+    if (this.vin == null) {
+      throw new IOException("InputStream is not initialized");
+    }
     byte[] magic = new byte[DataFileConstants.MAGIC.length];
     try {
       vin.readFixed(magic); // read magic
     } catch (IOException e) {
       throw new IOException("Not an Avro data file.", e);
     }
+    return magic;
+  }
+
+  void validateMagic(byte[] magic) throws InvalidAvroMagicException {
     if (!Arrays.equals(DataFileConstants.MAGIC, magic))
       throw new InvalidAvroMagicException("Not an Avro data file.");
+  }
+
+  /** Initialize the stream by reading from its head. */
+  void initializeWithMagic(InputStream in, byte[] magic) throws IOException {
```

Comment:
To my taste, I would have left this function named `initialize`. The reason is there is a pattern followed throughout to call the non-constructor initializers `initialize()`. But I wouldn't insist on it.

Reply:
Thanks @thiru-mg . I have fixed this in recent commit.